web/packages/teleterm/src/ui/services/modals/modalsService.ts

```diff
@@ -58,20 +67,44 @@ export class ModalsService extends ImmutableStore<State> {
    * old dialog with the new one.
    * The old dialog is canceled, if possible.
    *
-   * The returned closeDialog function can be used to close the dialog and automatically call the
-   * dialog's onCancel callback (if present).
+   * The passed `abortSignal` or the returned `closeDialog` function can be used to close the dialog
+   * and automatically call the dialog's onCancel callback (if present).
    */
-  openRegularDialog(dialog: Dialog): { closeDialog: () => void } {
+  openRegularDialog(
+    dialog: Dialog,
+    abortSignal?: AbortSignal
+  ): {
+    closeDialog: () => void;
+  } {
+    const onCancelDialog = () => dialog['onCancel']?.();
+    const sharedSignal = this.getSharedAbortSignal(abortSignal);
+    if (sharedSignal.aborted) {
+      onCancelDialog();
+      return {
+        closeDialog: () => {},
+      };
+    }
+
     this.state.regular?.['onCancel']?.();
```

Comment:
Let's change this to something like:

```
const previousDialog = this.state.regular
previousDialog?.['onCancel']?.()
```

Otherwise it gets kind of confusing seeing `dialog['onCancel']?.()` and `this.state.regular?.['onCancel']?.()` next to each other.